dali/operators/reader/loader/numpy_loader.cc

```diff
@@ -38,6 +37,115 @@ TypeInfo TypeFromNumpyStr(const std::string &format) {
   DALI_FAIL("Unknown Numpy type string");
 }
 
+inline void SkipSpace(const char*& ptr) {
+  while(::isspace(*ptr))
+    ptr++;
+}
+
+template <size_t N>
+void Skip(const char*& ptr, const char (&what)[N]) {
+  DALI_ENFORCE(!strncmp(ptr, what, N-1), make_string("Expected \"", what, "\" but got \"", ptr, "\""));
+  ptr += N - 1;
+}
+
+template <size_t N>
+bool TrySkip(const char*& ptr, const char (&what)[N]) {
+  if (!strncmp(ptr, what, N-1)) {
+    ptr += N - 1;
+    return true;
+  } else {
+    return false;
+  }
+}
+
+
+template <size_t N>
+void SkipFieldName(const char*& ptr, const char (&name)[N]) {
+  SkipSpace(ptr);
+  Skip(ptr, "'");
+  Skip(ptr, name);
+  Skip(ptr, "'");
+  SkipSpace(ptr);
+  Skip(ptr, ":");
+  SkipSpace(ptr);
+}
+
+std::string ParseStringValue(const char*& input, char delim_start = '\'', char delim_end = '\'') {
+  DALI_ENFORCE(*input++ == delim_start, make_string("Expected \'", delim_start, "\'"));
+  std::string out;
+  for (;*input != '\0'; input++) {
+    if (*input == '\\') {
+      switch (*++input) {
+        case '\'':
+          out += '\\';
+          break;
+        case '\t':
+          out += '\t';
+          break;
+        case '\n':
+          out += '\n';
+          break;
+        case '\"':
+          out += '\"';
+          break;
+        default:
+          out += '\\';
+          out += *input;
+          break;
+      }
+    } else if (*input == delim_end) {
+      break;
+    } else {
+      out += *input;
+    }
+  }
+  DALI_ENFORCE(*input++ == delim_end, make_string("Expected \'", delim_end, "\'"));
+  return out;
+}
+
+void ParseHeaderMetadata(NumpyParseTarget& target, std::string header) {
+  const char* hdr = header.c_str();
+  SkipSpace(hdr);
+  Skip(hdr, "{");
+  SkipFieldName(hdr, "descr");
+  auto typestr = ParseStringValue(hdr);
+    // < means LE, | means N/A, = means native. In all those cases, we can read
+  bool little_endian = (typestr[0] == '<' || typestr[0] == '|' || typestr[0] == '=');
+  DALI_ENFORCE(little_endian, "Big Endian files are not supported.");
+  target.type_info = TypeFromNumpyStr(typestr.substr(1));
+
+  SkipSpace(hdr);
+  Skip(hdr, ",");
+  SkipFieldName(hdr, "fortran_order");
+  if (TrySkip(hdr, "True")) {
+    target.fortran_order = true;
+  } else if (TrySkip(hdr, "False")) {
+    target.fortran_order = false;
+  } else {
+    DALI_FAIL("Can not parse fortran order");
+  }
+  SkipSpace(hdr);
+  Skip(hdr, ",");
+  SkipFieldName(hdr, "shape");
+  auto sh_str = ParseStringValue(hdr, '(', ')');
+  auto shape_values = string_split(sh_str, ',');
+  target.shape.reserve(shape_values.size());
+  try {
+    for (const auto &s : shape_values) {
+      if (s.empty())
+        continue;
+      target.shape.push_back(static_cast<int64_t>(stoi(s)));
+    }
+  } catch (...) {
+    DALI_FAIL(make_string("Failed to parse shape data: ", sh_str));
+  }
```

Comment:
How about catching `std::exception` and adding error msg to the error msg?
```c++
} catch(std::exception &e) {
  DALI_FAIL(make_string("Failed to parse shape data: ", sh_str, ". Error: ", e.what()));
}
```

Reply:
According to cppreference, `std::stoi` can throw only erros derived from logic_error (out_of_range and invalid_argument), so that's what we should catch.

Reply:
done